x_transformers/attend.py

```diff
@@ -137,8 +137,10 @@ def flash_attn(
             # manually handle causal mask, if another mask was given
 
             if causal:
-                causal_mask = torch.ones((q_len, k_len), dtype = torch.bool, device = device).triu(k_len - q_len + 1)
-                mask = mask & ~causal_mask
+                r = torch.arange(q_len, device=device)
+                causal_mask = rearrange(r, "q_len -> () () q_len ()") < rearrange(r, "k_len -> () () () k_len")
+                causal_mask = F.pad(causal_mask, (k_len - q_len, 0), value=False)
+                mask = mask | causal_mask
```

Comment:
i thought `mask` here was `True` for keep, `False` to mask out (and causal mask is `True` to mask out)

Reply:
You're correct! Thanks for catching this and shows I need to be more careful in preparing a PR. The repo where I originally tested, I had the right code but creating the fork and copying the changes I made this mistake. Sorry about it.